src/Microsoft.DotNet.Darc/tests/Microsoft.DotNet.Darc.VirtualMonoRepo.E2E.Tests/VmrTests.cs

```diff
@@ -0,0 +1,376 @@
+    private string? submodulePath;
+    private string? vmrPath;
+    private string? tmpPath;
+    private readonly string baseDir;
+    private readonly ProcessManager processManager;
+    private readonly string darcExecutable;
+
+    public VmrTests()
+    {
+        processManager = new ProcessManager(new NullLogger<ProcessManager>(), "git");
+        var assembly = Assembly.GetAssembly(typeof(VmrTests));
+        if(assembly == null)
+        {
+            throw new Exception("Assembly not found");
+        }
+
+        darcExecutable = Path.Combine(assembly.Location, "..", "Microsoft.DotNet.Darc.exe");
+        baseDir = Path.Combine(Path.GetTempPath(), "_vmrTests");
+    }
+
+    [SetUp]
+    public async Task SetUp()
+    {
+        tempDir = Path.Combine(baseDir, Path.GetRandomFileName());
+        Directory.CreateDirectory(tempDir);
+
+        testRepoPath = Path.Combine(tempDir, "test-repo");
+        externalRepoPath = Path.Combine(tempDir, "external-repo");
+        submodulePath = Path.Combine(tempDir, "test-repo", "externals", "external-repo");
+        vmrPath = Path.Combine(tempDir, "vmr");
+        tmpPath = Path.Combine(tempDir, "tmp");
+
+        Directory.CreateDirectory(vmrPath);
+        Directory.CreateDirectory(testRepoPath);
+        Directory.CreateDirectory(Path.Combine(testRepoPath, "excluded"));
+        Directory.CreateDirectory(tmpPath);
+        Directory.CreateDirectory(externalRepoPath);
+
+        var sourceMappings = string.Format(@"{{
+            ""patchesPath"": ""src/installer/src/SourceBuild/tarball/patches"",
+            ""defaults"": {{
+              ""defaultRef"": ""main"",
+              ""exclude"": [
+                ""**/*.dll""
+              ]
+            }},
+            ""mappings"": [
+              {{
+                ""name"": ""test-repo"",
+                ""defaultRemote"": ""{0}"",
+                ""exclude"": [
+                  ""externals/external-repo/**/*.exe"",
+                  ""excluded/*""
+                ]
+              }}
+            ]
+        }}", testRepoPath.Replace("\\", "\\\\"));
+
+        File.WriteAllText(Path.Combine(testRepoPath, "test-repo-file.txt"), "Test repo file");
+        File.WriteAllText(Path.Combine(testRepoPath, "excluded", "excluded.txt"), "File to be excluded");
+
+        Directory.CreateDirectory(Path.Combine(vmrPath, "src"));
+        File.WriteAllText(Path.Combine(vmrPath, "src", "source-mappings.json"), sourceMappings);
+
+        await InitialCommit(vmrPath);
+        await InitialCommit(testRepoPath);
+    }
+
+    [TearDown]
+    public void CleanUpOutputFile()
+    {
+        try
+        {
+            if (tempDir is not null)
+            {
+                DeleteDirectory(tempDir);
+            }
+        }
+        catch
+        {
+            // Ignore
+        }
+    }
+
+    [Test]
+    public async Task RepoIsInitializedTest()
+    {
+        if (testRepoPath == null || vmrPath == null || tmpPath == null)
```

Comment:
I think it's better to make these non-nullable and initialize them with `null!` because you know that they will be set by the testing framework. Then you don't need these checks